packages/driver/src/multidomain/index.js

```diff
@@ -80,11 +73,91 @@ const setup = () => {
   Cypress.on('log:added', onLogAdded)
   Cypress.on('log:changed', onLogChanged)
 
-  specBridgeCommunicator.on('run:domain:fn', ({ data, fn }) => {
-    // TODO: await this if it's a promise, or do whatever cy.then does
-    window.eval(`(${fn})`)(data)
+  const doneEarly = () => {
+    cy.queue.stop()
+
+    // we only need to worry about doneEarly when
+    // it comes from a manual event such as stopping
+    // Cypress or when we yield a (done) callback
+    // and could arbitrarily call it whenever we want
+    const p = cy.state('promise')
+
+    // if our outer promise is pending
+    // then cancel outer and inner
+    // and set canceled to be true
+    if (p && p.isPending()) {
+      cy.state('canceled', true)
+      cy.state('cancel')()
+    }
+
+    // if a command fails then after each commands
+    // could also fail unless we clear this out
+    cy.state('commandIntermediateValue', undefined)
+
+    // reset the nestedIndex back to null
+    cy.state('nestedIndex', null)
+  }
+
+  specBridgeCommunicator.on('run:domain:fn', async ({ data, fn, isDoneFnAvailable = false }) => {
+    cy.reset({})
+
+    cy.state('runnable', {
+      ctx: {},
+      clearTimeout () {},
+      resetTimeout () {},
+      timeout () {},
+    })
+
+    let fnWrapper = `(${fn})`
 
-    specBridgeCommunicator.toPrimary('ran:domain:fn')
+    if (isDoneFnAvailable) {
+      // stub out the 'done' function if available in the primary domain
+      // to notify the primary domain if the done() callback is invoked
+      // within the spec bridge
+      const done = (err = undefined) => {
+        // TODO: calling this currently causes some queue issues with subsequent tests. This needs to be explored as to why in the near future
```

Comment:
This comment can be removed, right? I think we handled this by implementing the done signal (user passing in done function to `switchToDomain`).

Reply:
It can! I think this one was pertaining to the `doneEarly` function, which we wound up implementing for the secondary domain. Should be axed out in f670f24b4ea77652022846a330fd2002ea9f29ce